mitiq/rem/rem.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (C) 2021 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Readout Confusion Inversion."""
+
+from typing import Callable, Optional, Union
+
+from functools import wraps
+import numpy as np
+
+from mitiq._typing import QPROGRAM, MeasurementResult
+from mitiq.executor.executor import Executor
+from mitiq.observable.observable import Observable
+from mitiq.rem.inverse_confusion_matrix import mitigate_measurements
+
+
+def execute_with_rem(
+    circuit: QPROGRAM,
+    executor: Union[Executor, Callable[[QPROGRAM], MeasurementResult]],
+    observable: Optional[Observable] = None,
+    *,
+    inverse_confusion_matrix: np.ndarray,
+) -> float:
+    """Returns the readout error mitigated expectation value utilizing an
+    inverse confusion matrix.
+
+    Args:
+        executor: A Mitiq executor that executes a circuit and returns the
+            unmitigated ``MeasurementResult``.
+        observable: Observable to compute the expectation value of.
+        inverse_confusion_matrix: The inverse confusion matrix to apply to the
+            probability vector estimated with noisy measurement results.
+    """
+    if not isinstance(executor, Executor):
+        executor = Executor(executor)
+
+    result = executor._run([circuit])
+    noisy_result = result[0]
+    assert isinstance(noisy_result, MeasurementResult)
+
+    mitigated_result = mitigate_measurements(
+        noisy_result, inverse_confusion_matrix
+    )
+    assert observable is not None
+    return observable._expectation_from_measurements([mitigated_result])
```

Comment:
```suggestion
    noisy_result = executor._run([circuit])[0]
    assert isinstance(noisy_result, MeasurementResult)

    mitigated_result = mitigate_measurements(
        noisy_result, inverse_confusion_matrix
    )
    assert observable is not None
    return observable._expectation_from_measurements([mitigated_result])
```

We don't have many `assert` statements in our code outside of tests. Is there a reason to use them here over more explicitly raising an error?

Reply:
I've replaced these with exceptions